File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/FertiliserStore.kt
package com.rork.vinetrack.data

import android.content.Context
import androidx.core.content.edit
import com.rork.vinetrack.data.model.FertiliserRecord
import kotlinx.serialization.json.Json

/**
 * Offline-first local cache for the Fertiliser Calculator (in development —
 * System Admin only). Backs saved calculations with per-vineyard JSON blobs in
 * SharedPreferences. This is the CACHE layer for the shared
 * `fertiliser_records` / `fertiliser_record_allocations` Supabase tables —
 * [FertiliserSyncCoordinator] writes through it and reconciles it with the
 * server.
 *
 * Products are NOT cached here — the Fertiliser Calculator reads its product
 * library from the shared saved chemical database (`saved_chemicals`,
 * sql/111), which lives in the main app state.
 */
class FertiliserStore(context: Context) {

    private val prefs = context.applicationContext
        .getSharedPreferences("vinetrack_fertiliser", Context.MODE_PRIVATE)

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    fun loadRecords(vineyardId: String): List<FertiliserRecord> {
        val raw = prefs.getString("records_v2_$vineyardId", null) ?: return emptyList()
        return runCatching { json.decodeFromString<List<FertiliserRecord>>(raw) }
            .getOrDefault(emptyList())
            .sortedByDescending { it.date }
    }

    fun saveRecords(vineyardId: String, records: List<FertiliserRecord>) {
        prefs.edit { putString("records_v2_$vineyardId", json.encodeToString(records)) }
    }

    fun addRecord(vineyardId: String, record: FertiliserRecord): List<FertiliserRecord> {
        val updated = loadRecords(vineyardId) + record
        saveRecords(vineyardId, updated)
        return updated.sortedByDescending { it.date }
    }

    fun markCompleted(vineyardId: String, recordId: String, date: String): List<FertiliserRecord> {
        val updated = loadRecords(vineyardId).map {
            if (it.id == recordId) it.copy(status = "completed", date = date) else it
        }
        saveRecords(vineyardId, updated)
        return updated
    }

    fun deleteRecord(vineyardId: String, recordId: String): List<FertiliserRecord> {
        val updated = loadRecords(vineyardId).filterNot { it.id == recordId }
        saveRecords(vineyardId, updated)
        return updated
    }
}

File: ios/VineTrackV2/LegacyImported/Services/FertiliserStore.swift
import Foundation
import Observation

/// Offline-first store for the Fertiliser Calculator (System Admin only while
/// in development). Local cache for the shared `fertiliser_records` /
/// `fertiliser_record_allocations` Supabase tables: writes land here first and
/// fire sync hooks; remote state is applied through the `applyRemote*`
/// methods, which never re-fire the hooks.
///
/// Products are NOT stored here — the Fertiliser Calculator reads its product
/// library from the shared saved chemical database (`MigratedDataStore.savedChemicals`,
/// sql/111), which has its own sync service.
@Observable
final class FertiliserStore {
    static let shared = FertiliserStore()

    private(set) var records: [FertiliserRecord] = []

    /// Sync hooks — fired for local user edits only, never for remote applies.
    var onRecordChanged: ((UUID) -> Void)?
    var onRecordDeleted: ((UUID) -> Void)?

    private static let recordsKey = "vinetrack_fertiliser_records_v2"

    private let persistence: PersistenceStore

    init(persistence: PersistenceStore = .shared) {
        self.persistence = persistence
        records = persistence.load(key: Self.recordsKey) ?? []
    }

    // MARK: Records

    func records(forVineyard vineyardId: UUID?) -> [FertiliserRecord] {
        let list = vineyardId.map { id in records.filter { $0.vineyardId == id } } ?? records
        return list.sorted { $0.date > $1.date }
    }

    func addRecord(_ record: FertiliserRecord) {
        applyRecordUpsert(record)
        persistRecords()
        onRecordChanged?(record.id)
    }

    /// Converts a planned task into a completed application record.
    func markCompleted(id: UUID, on date: Date = Date()) {
        guard let index = records.firstIndex(where: { $0.id == id }) else { return }
        records[index].status = .completed
        records[index].date = date
        persistRecords()
        onRecordChanged?(id)
    }

    func deleteRecord(id: UUID) {
        records.removeAll { $0.id == id }
        persistRecords()
        onRecordDeleted?(id)
    }

    // MARK: Remote applies (no hooks)

    func applyRemoteRecordUpsert(_ record: FertiliserRecord) {
        applyRecordUpsert(record)
        persistRecords()
    }

    func applyRemoteRecordDelete(_ id: UUID) {
        let before = records.count
        records.removeAll { $0.id == id }
        if records.count != before { persistRecords() }
    }

    // MARK: Persistence

    private func applyRecordUpsert(_ record: FertiliserRecord) {
        if let index = records.firstIndex(where: { $0.id == record.id }) {
            records[index] = record
        } else {
            records.append(record)
        }
    }

    private func persistRecords() {
        persistence.save(records, key: Self.recordsKey)
    }
}
